### PR TITLE
Detect and stop sign-ins refused by Omada's logon page

### PR DESCRIPTION
**Pull request by mvaneijken**
## Summary

Add detection and handling of sign-in failures that occur when the identity provider redirects back to Omada's logon page with an error message. Previously, the module would retry indefinitely when a federated sign-in was refused (e.g., account not in tenant), resulting in half an hour of repeated browser windows landing on the same error page.

## Key Changes

- **New error detection script** (`Get-OmadaLogonErrorScript`): JavaScript that scrapes Omada's logon page to extract error messages, form visibility, and page context. Shared by both WebView2 and Edge WebDriver drivers to ensure consistent behavior.

- **New error classification logic** (`Test-OmadaLogonPageError`): Determines whether an error is terminal (cannot be retried) or correctable (user can act on it). Terminal errors include:
  - Identity provider failures with AADSTS codes or OpenID Connect error responses
  - Account/tenant membership issues
  - Any error on a logon page with no visible credential fields to retry with

- **WebView2 integration** (`Get-WebView2LogonPageError`): Asynchronously reads the logon page within the 150ms timer tick, throttles scrapes to avoid excessive polling, and stops the sign-in for terminal errors while leaving correctable errors (like wrong password) for the user to fix.

- **Edge WebDriver integration** (`Get-WebDriverLogonPageError`): Synchronous equivalent for the Edge WebDriver path, sharing the same error classification logic.

- **Sign-in termination** (`Stop-OmadaLogin`): Records the refusal reason in `$Script:LoginAbortReason` so both drivers stop retrying, disables credential autofill, and reports the error with code, URL, and explanation to the user.

- **Driver updates**: Modified `Get-DataFromWebView2` and `Get-DataFromWebDriver` to check for recorded abort reasons before opening new browser windows, and to report the page's own error message instead of generic authentication failures.

## Implementation Details

- Error detection runs asynchronously in WebView2 (result picked up on next timer tick) to avoid blocking the UI thread
- Scrapes are throttled to prevent excessive polling of pages that are merely loading
- Correctable errors (wrong password on Omada's form) are reported once and otherwise left alone
- Terminal errors immediately close the browser and prevent further retry attempts
- All error messages are sanitized through `Protect-LogMessage` before logging to avoid exposing sensitive request data
- Comprehensive unit tests cover both drivers, error classification rules, and edge cases (closed browsers, navigation interruptions, etc.)

https://claude.ai/code/session_01Keceeq2Y3irvzzhzj1AJTr

### CODE REVIEW

File: OmadaWeb.PS/Private/Get-WebView2LogonPageError.ps1 [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    # The watchdog counters belong to the window that is closing, not to the next one.
+    $Script:OmadaWatchdogStart = $null
+    $Script:OmadaWatchdogRunning = $false
+    $Script:ProgressCounter = 0
+    $Script:LastFiredSecond = -1
+
+    if ($null -ne $Script:WebView2 -and $null -ne $Script:WebView2.FindForm()) {
+        $Script:WebView2.FindForm().Close()
+    }
```
[the remark this answers is not in the document]
Comment (mvaneijken) [reply OmadaWeb.PS/Private/Get-WebView2LogonPageError.ps1]: Fixed in 75ad1ca. The form is now resolved once into $LoginForm, null-checked, and closed.

File: OmadaWeb.PS/Private/Get-OmadaLogonErrorScript.ps1 [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        var wrapsAnother = false;
+        for (var j = 0; j < candidates.length; j++) {
+            if (i !== j && candidate.contains(candidates[j])) { wrapsAnother = true; break; }
+        }
+        if (wrapsAnother) { continue; }
+        if (!isVisible(candidate)) { continue; }
+
+        var message = readText(candidate);
+        if (!message || message.length > 4000) { continue; }
+        if (messages.indexOf(message) !== -1) { continue; }
```
[the remark this answers is not in the document]
Comment (mvaneijken) [reply OmadaWeb.PS/Private/Get-OmadaLogonErrorScript.ps1:119]: Good catch, and the consequence was the one you name: a long banner made the page read as having no error, so the retry-stop never fired. Fixed in 75ad1ca - the message is truncated to 4000 characters instead of being discarded. Test-OmadaLogonPageError still truncates to 2000 for display, so the cap here only keeps the payload out of the megabyte range.

File: OmadaWeb.PS/Private/Stop-OmadaLogin.ps1 [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        return $false
+    }
+
+    $PageAddress = "unknown"
+    if (-not [string]::IsNullOrWhiteSpace($Url)) {
+        try {
+            $PageAddress = ([System.Uri]::New($Url)).GetLeftPart([System.UriPartial]::Path)
+        }
+        catch {
+            $PageAddress = Protect-LogMessage -Message $Url
```
[the remark this answers is not in the document]
Comment (mvaneijken) [reply OmadaWeb.PS/Private/Stop-OmadaLogin.ps1]: Fixed in 75ad1ca. The catch path now cuts the string at the first '?' or '#' before redacting, so a URL that will not parse as a Uri cannot put query or fragment data into the warning.
